asmcomp/cmmgen.ml

```diff
@@ -224,6 +224,112 @@ let untag_int = function
   | Cop(Cor, [c; Cconst_int 1]) -> Cop(Casr, [c; Cconst_int 1])
   | c -> Cop(Casr, [c; Cconst_int 1])
 
+let int_comp_with_constant cmp arg n =
+  let default = Cop(Ccmpi cmp, [arg; Cconst_int n]) in
+  match arg with
+  | Cop(Cor, [Cop(Clsr, [arg; Cconst_int shift]); Cconst_int mask])
+  | Cop(Cor, [Cconst_int mask; Cop(Clsr, [arg; Cconst_int shift])]) ->
+      (* Optimize the pattern:
+           ( (a >> shift) | mask )  <  n
+         into
+           ( a & ( !mask << shift ) ) < (n & !mask) << shift
+         when
+           A: n | mask = n
+           B: (n & !mask) << shift >> shift = (n & !mask)
+           C: sign ((n & !mask) << shift) = sign n
+
+         shift, mask and n being constants, the remaining operations are:
+            a & and_mask < n'
+         which is one less operation
+
+         =======
+
+         For all comparison operator (<)
+
+                              ( (a >> shift) | mask )  <  n
+
+         iff
+
+                              ( (a >> shift) | mask )  <  (n | mask)
```

Comment:
Effectively. I missed thought about some problems on sign overflows on n but I don't know why I forgot on a.
By the way there is another similar one on n. The constraint B is a bit too weak for preventing sign problems on n:
the second step does not hold if mask is negative.

Well since it is not possible to test the right constraints on a not to have problems with the signs, and everything being a lot simpler with unsigned comparison. I think I will try in selectgen as you suggested.

And z3 seems not to complain about it:
The sign constraint being replaced by simply requiring that shift is not null (to ensure that changing the test from signed to unsigned does not change anything)
```
(set-logic QF_BV)

(declare-fun a () (_ BitVec 32))
(declare-fun shift () (_ BitVec 32))
(declare-fun mask () (_ BitVec 32))
(declare-fun n () (_ BitVec 32))

(assert (= (bvor n mask) n))
(assert (= (bvlshr (bvshl (bvand n (bvnot mask)) shift) shift) (bvand n (bvnot mask))))
(assert (not (= #x00000000  shift)))

(assert (not (= (bvslt (bvor (bvlshr a shift) mask) n)
                (bvult (bvand a (bvshl (bvnot mask) shift)) (bvshl (bvand n (bvnot mask)) shift))
         )))

(check-sat)
(exit)
```

Thank you for taking time to find that.